src/hotspot/share/opto/divnode.cpp

```diff
@@ -852,31 +1088,80 @@ Node* UDivINode::Identity(PhaseGVN* phase) {
 // prevent hoisting the divide above an unsafe test.
 const Type* UDivINode::Value(PhaseGVN* phase) const {
   // Either input is TOP ==> the result is TOP
-  const Type *t1 = phase->type( in(1) );
-  const Type *t2 = phase->type( in(2) );
-  if( t1 == Type::TOP ) return Type::TOP;
-  if( t2 == Type::TOP ) return Type::TOP;
+  const Type* t1 = phase->type(in(1));
+  const Type* t2 = phase->type(in(2));
+  if(t1 == Type::TOP || t2 == Type::TOP) {
+    return Type::TOP;
+  }
 
   // x/x == 1 since we always generate the dynamic divisor check for 0.
   if (in(1) == in(2)) {
     return TypeInt::ONE;
   }
 
-  // Either input is BOTTOM ==> the result is the local BOTTOM
-  const Type *bot = bottom_type();
-  if( (t1 == bot) || (t2 == bot) ||
-      (t1 == Type::BOTTOM) || (t2 == Type::BOTTOM) )
-    return bot;
+  // TODO: Improve Value inference of both signed and unsigned division
+  const TypeInt* i1 = t1->isa_int();
+  const TypeInt* i2 = t2->isa_int();
+  assert(i1 != nullptr && i2 != nullptr, "");
+  if (i1->is_con() && i2->is_con()) {
+    return TypeInt::make(juint(i1->get_con()) / juint(i2->get_con()));
+  }
 
   // Otherwise we give up all hope
   return TypeInt::INT;
 }
 
 //------------------------------Idealize---------------------------------------
-Node *UDivINode::Ideal(PhaseGVN *phase, bool can_reshape) {
+Node* UDivINode::Ideal(PhaseGVN* phase, bool can_reshape) {
   // Check for dead control input
-  if (in(0) && remove_dead_region(phase, can_reshape))  return this;
-  return NULL;
+  if (in(0) && remove_dead_region(phase, can_reshape)) {
+    return this;
+  }
+  // Don't bother trying to transform a dead node
+  if(in(0) && in(0)->is_top()) {
+    return nullptr;
+  }
+
+  const Type* t = phase->type(in(2));
+  if(t == TypeInt::ONE) {       // Identity?
+    return nullptr;             // Skip it
+  }
+
+  const TypeInt *ti = t->isa_int();
+  if(ti == nullptr) {
+    return nullptr;
+  }
+
+  // Check for useless control input
+  // Check for excluding div-zero case
+  if (in(0) && (ti->_hi < 0 || ti->_lo > 0)) {
+    set_req(0, nullptr);           // Yank control input
+    return this;
+  }
+
+  // Divisor very large, constant 2**31 can be transform to a shift
```

Comment:
@vnkozlov Oh the transformation is here

Reply:
It only works for cases with narrowed type. Which is not common case unless Java user's code check for that.

Reply:
May you expand on this, please? How can I know that a divisor is negative without using its type information? Thanks.

Reply:
What I am trying to say is that you can not replace removed runtime code in `C2_MacroAssembler` with only this code in IR (which covers only part of cases). You need additional code for case when divisor is not constant and its type is unknown. You need to generate IR code similar to runtime code with explicit runtime check for negative divisor and calculation `(dividend & ~(dividend - divisor)) >>> (Integer.SIZE - 1)`. Which will be very disruptive for IR graph.

My suggestion is to leave this your code for cases when type is known - it will allow to avoid `div` instruction generation and it is good. But also restore `C2_MacroAssembler` code you removed - it will cover the rest of cases when type (negativity) of divisor is not known during compilation.

Reply:
Thanks for your explanation. I believe this can be achieved by changing the name of the intrinsic to `divideUnsigned0` and implementing `Integer/Long::divideUnsigned` as follow:

    int divideUnsigned(int dividend, int divisor) {
        if (divisor > 0) {
            return divideUnsigned0(dividend, divisor);
        } else if (divisor < 0) {
            return compareUnsigned(dividend, divisor) >= 0 ? 1 : 0;
        } else {
            throw new ArithmeticException();
        }
    }

However, I think that this is premature optimisation as:

- Integer values are heavily biased toward 0, in unsigned contexts, a negative divisor is a really large divisor, which is highly unlikely. As a result, this results in redundant code for the majority of cases.
- This optimisation can be implemented purely in Java, which means that the users of this method can implement it themselves if it is known that the distribution of divisors is different from the general cases.

If you don't agree then I would implement the optimisation in Java as presented above. Thanks very much.

Reply:
The question is: do we complicate the code for very rare case (negative variable devisor with unknown type during compilation) or we slightly improve common case (by remove check for negative devisor at runtime) but introducing regressing for that rare case?

May be this discussion is mute of regression is small. Please, get data for negative divisor in your JMH test before we decide what to do.

I assume `divl` instruction produces correct result for negative divisor - we will be using it now with your changes.

Reply:
The JMH results and the source are below. The overhead of a branch in the common path is less noticeable in the cases of long divisions where the cost of the division is much larger. However, my CPU is based on Skylake architecture with really expensive `divq` instruction, and according to my knowledge, other architectures (Intel from Icelake, AMD Zen, and maybe AArch) may have a `divq` with a cost similar to a `divl`.

                                                                                       Baseline           This patch
    Benchmark                                  (BUFFER_SIZE)  (type)  Mode  Cnt     Score     Error     Score     Error  Units
    IntegerDivMod.testDivideUnsignedNaive               1024     pos  avgt   25  2549.410 ±  45.532  2162.241 ±  38.560  ns/op
    IntegerDivMod.testDivideUnsignedNaive               1024     neg  avgt   25  1115.032 ±  33.985  2176.231 ±  23.436  ns/op
    IntegerDivMod.testDivideUnsignedNaive               1024     mix  avgt   25  1814.831 ±  28.754  2185.103 ±  47.785  ns/op
    IntegerDivMod.testDivideUnsignedOptimised           1024     pos  avgt   25  2703.873 ±  47.586  2345.205 ±  35.003  ns/op
    IntegerDivMod.testDivideUnsignedOptimised           1024     neg  avgt   25   926.830 ±  16.540   922.096 ±  17.555  ns/op
    IntegerDivMod.testDivideUnsignedOptimised           1024     mix  avgt   25  1877.774 ± 204.858  1547.250 ± 120.576  ns/op
    LongDivMod.testDivideUnsignedNaive                  1024     pos  avgt   25  7245.008 ±  92.670  7239.190 ±  67.525  ns/op
    LongDivMod.testDivideUnsignedNaive                  1024     neg  avgt   25   987.773 ±  13.338  7176.254 ±  68.509  ns/op
    LongDivMod.testDivideUnsignedNaive                  1024     mix  avgt   25  4537.198 ±  63.926  7189.320 ±  58.801  ns/op
    LongDivMod.testDivideUnsignedOptimised              1024     pos  avgt   25  7787.518 ±  74.144  7379.395 ±  63.933  ns/op
    LongDivMod.testDivideUnsignedOptimised              1024     neg  avgt   25   617.233 ±   8.030   615.633 ±   5.545  ns/op
    LongDivMod.testDivideUnsignedOptimised              1024     mix  avgt   25  4532.729 ±  57.028  4440.418 ±  42.612  ns/op

The benchmark is derived from the existing ones:

    public class IntegerDivMod {

        RandomGenerator randomGenerator;

        @Param({"pos", "neg", "mix"})
        String type;

        @Param({"1024"})
        int BUFFER_SIZE;
        int[] dividends, divisors, quotients, remainders;

        @Setup
        public void setup() {
            dividends = new int[BUFFER_SIZE];
            divisors = new int[BUFFER_SIZE];
            quotients =  new int[BUFFER_SIZE];
            remainders =  new int[BUFFER_SIZE];
            var rng = RandomGenerator.getDefault();
            for (int i = 0; i < BUFFER_SIZE; i++) {
                dividends[i] = rng.nextInt();
                int divisor = rng.nextInt();
                divisor = divisor == 0 ? 1 : divisor;
                if (type.equals("pos")) {
                    if (divisor == Integer.MIN_VALUE) {
                        divisor = Integer.MAX_VALUE;
                    } else if (divisor < 0) {
                        divisor = -divisor;
                    }
                } else if (type.equals("neg")) {
                    if (divisor > 0) {
                        divisor = -divisor;
                    }
                }
                divisors[i] = divisor;
            }
        }

        @Benchmark
        public void testDivideUnsignedNaive() {
            for (int i = 0; i < BUFFER_SIZE; i++) {
                quotients[i] = Integer.divideUnsigned(dividends[i], divisors[i]);
            }
        }

        @Benchmark
        public void testDivideUnsignedOptimised() {
            for (int i = 0; i < BUFFER_SIZE; i++) {
                int divisor = divisors[i];
                if (divisor < 0) {
                    quotients[i] = Integer.compareUnsigned(dividends[i], divisor) >= 0 ? 1 : 0;
                } else if (divisor > 0) {
                    quotients[i] = Integer.divideUnsigned(dividends[i], divisor);
                } else {
                    throw new ArithmeticException();
                }
            }
        }
    }

Reply:
Even ignoring Long case the difference is almost 2x for Integer negative case vs ~15% improvement for positive. It is not good tradeoff.

RFE is filed for constants. I don't think you need to touch variable case.